src/image.rs

```diff
@@ -0,0 +1,198 @@
+// This function is able to parse a variety of files and returns a byte array that represents a pico-8 readable format.
+// Supported formats are:
+// - .p8
+// - .p8.png
+// - .png: assumes dimensions divisible by 128 (screenshot)
+
+// load_image:w,h,frames,mode
+// w = number between 1 and 128
+// w = number between 1 and 128
+// mode: # In the future there will be different drawing modes that allow you to display the image with scanlines.
+//         Because it is technically possible to screenshot up to 32 colors on the screen at one time.
+// - default: default palette mode. all colors will be converted/mapped to the default palette.
+// - custom:  custom palette mode, only 1 palette allowed.
+// - cust32:  outputs screen orientation (horizontal/vertical). then palette mappings for all rows. then each row.
+//
+
+// a 64x64 image can load at 60 fps if on max cpu. 30 fps if on lower cpu.
+
+// 1 second gif at 60 fps can be loaded in 4 seconds without any compression. though scan lines could be cycled and cached to guarantee max load speed.
+
+// - for 32 color gifs, up to 12 frames can be loaded a second (on high cpu),   6 frames a second to be safe.
+// - frames can in 4 frames on near max cpu (15 fps). 8 frames with half cpu (7.5 fps).
+// - palette can change each frame, so each frame is loading a separate image. if loading scanlines, there could be 2 bits before each scanline. 1 bit: palette, 2 bit: same line
+// - if switching rotation, not same as previous frame.
+
+// - gif playback could be sped up by checking scan lines that changed from the previous frame. but do i really want to do that?
+// gifs files only support 100fps, 50fps, 33.3fps, 25fps, and slower.
+
+// frames can be loaded 10 frames can be loaded in 1 se
+
+use image::{DynamicImage, RgbaImage};
+use lazy_static::lazy_static;
+use ndarray::{arr1, arr2, Array1, Array2};
+use tokio::{fs, task};
+use std::cmp::Ordering;
+use std::env;
+use std::path::Path;
+
+lazy_static! {
+    static ref DEFAULT_PALETTE: Array2<u8> = arr2(&[
+        [0, 0, 0],
+        [29, 43, 83],
+        [126, 37, 83],
+        [0, 135, 81],
+        [171, 82, 54],
+        [95, 87, 79],
+        [194, 195, 199],
+        [255, 241, 232],
+        [255, 0, 77],
+        [255, 163, 0],
+        [255, 236, 39],
+        [0, 228, 54],
+        [41, 173, 255],
+        [131, 118, 156],
+        [255, 119, 168],
+        [255, 204, 170]
+    ]);
+    static ref EXTENDED_PALETTE: Array2<u8> = arr2(&[
```

Comment:
Label images can actually have 32 colors. They can have extended palette colors or normal palette colors. Goal is to have a protocol that supports any colored image. The protocol could send the palette colors to pico8, and pico8 could change the screen palette to show the correct colors for the image.

Technically a label image could have an impossible image in pico8, because pico8 has a scanline limitation on the palettes, but the label does not. So it would be cool to have a protocol that does a best approximation for this extreme edge case.